### PR TITLE
XArapucas Updates

### PR DESCRIPTION
**Pull request by rodralva**
XArapuca Updates in the OpDetSim Module related to issues [#78](https://github.com/SBNSoftware/sbndcode/issues/78), [#80](https://github.com/SBNSoftware/sbndcode/issues/80) :

-Added new sampling field to the sbnd_pds_mapping.json file.
-Modified digitizer, op hit finder, wvf analyzer and trigger modules to differentiate daphne xarapuca channels with their own frequency. 
-Updated test bench XArapuca single electron response.

The data driven response is loaded from sbnd_data  ([OpDetSim/digi_arapuca_sbnd.root](https://github.com/SBNSoftware/sbndcode/blob/feature/rodrigoa_XArapucasUpdates/sbndcode/OpDetSim/digi_arapuca_sbnd.fcl#L27)). The file needs to be updated adding the vectors with the single electron response data before the CI tests.

### CODE REVIEW

**Comment by ikatza** [comment]
I'm happy to take this one since I opened the issues.

**Comment by ikatza** [comment]
I believe this could use some refactoring. I've agreed to have a call with @rodralva to work on this, instead of writing all of my observations in here.

**Comment by wesketchum** [comment]
Assigned to upcoming production milestone, but I think if it needs more time it's ok to go. Also, further refactoring work can be done later if we think the code is OK now and it is important to start testing.

**Comment by rodralva** [comment]
I ran some test on the same G4 Event and printed the digitized waveforms with different configurations, everything looks fine. image

**Comment by pgreen135** [comment]
@rodralva is this PR now ready (pending CI test results)? if so @ikatza can you confirm that your concerns have been addressed / repeat the review?

**Comment by rodralva** [comment]
Yes, it is. For the CI test: I switched the [ideal response to default](https://github.com/SBNSoftware/sbndcode/blob/feature/rodrigoa_XArapucasUpdates/sbndcode/OpDetSim/digi_arapuca_sbnd.fcl#L28) (to use the test response, change the SinglePEmodel parameter to true). Also, keep in mind that the test response is loaded from sbnd_data 01_08_00 version.

File: sbndcode/OpDetSim/sbndPDMapAlg_tool.cc [hunk 1]
```diff
@@ -28,12 +28,23 @@ namespace opdet {
     return false;
   }
 
+  bool sbndPDMapAlg::isElectronics(size_t ch, std::string pdname) const
+  {
+    if(PDmap.at(ch)["electronics"] == std::string(pdname)) return true; // TODO: add number of electronics, daphne01, daphne02, .... ~rodrigoa; at() throws an exception if no such channel
```
Comment (ikatza) [inline sbndcode/OpDetSim/sbndPDMapAlg_tool.cc:33]: fine to leave the TODO, but the comment about `at()` is not necessary

File: sbndcode/OpDetSim/sbndPDMapAlg_tool.cc [hunk 2]
```diff
@@ -28,12 +28,23 @@ namespace opdet {
...
+      return false;
+}
+
   std::string sbndPDMapAlg::pdType(size_t ch) const
   {
-    if(ch < PDmap.size()) return PDmap.at(ch)["pd_type"];
+    if(ch < PDmap.size()) return PDmap.at(ch)["pd_type"];//TODO: remove all if(ch < PDmap.size()) ~rodrigoa
```
Comment (ikatza) [inline sbndcode/OpDetSim/sbndPDMapAlg_tool.cc:39]: the `if(ch < PDmap.size())` is not necessary

File: sbndcode/OpDetSim/sbndPDMapAlg_tool.cc [hunk 3]
```diff
@@ -28,12 +28,23 @@ namespace opdet {
...
     return "There is no such channel";
```
Comment (ikatza) [inline sbndcode/OpDetSim/sbndPDMapAlg_tool.cc:40]: Can you remove this return please, if it fails to find the channel is better to throw an exception from `at()`

File: sbndcode/OpDetSim/wvfAna_module.cc [hunk 4]
```diff
@@ -96,7 +99,7 @@ namespace opdet {
 
     auto const clockData = art::ServiceHandle<detinfo::DetectorClocksService const>()->DataForJob();
     fSampling = clockData.OpticalClock().Frequency(); // MHz
-
+    fSampling_Daphne=p.get<double >("DaphneFrequency" );
```
Comment (ikatza) [inline sbndcode/OpDetSim/wvfAna_module.cc:102]: Please, spaces around `=`, and remove the leading spaces between `>` and `)`

File: sbndcode/OpDetSim/wvfAna_module.cc [hunk 5]
```diff
@@ -160,7 +164,10 @@ namespace opdet {
                << "_" << hist_id;
 
       fStartTime = wvf.TimeStamp(); //in us
-      fEndTime = double(wvf.size()) / fSampling + fStartTime; //in us
+      if (opdetElectronics == "daphne"){
+		  fEndTime = double(wvf.size()) / fSampling_Daphne + fStartTime;} //in us
+			else{
```
Comment (ikatza) [inline sbndcode/OpDetSim/wvfAna_module.cc:169]: maybe is just GitHub rendering but indentation here seems wrong

File: sbndcode/OpDetSim/wvfAna_module.cc [hunk 6]
```diff
@@ -160,7 +164,10 @@ namespace opdet {
...
+      fEndTime = double(wvf.size()) / fSampling + fStartTime;} //in us
```
Comment (ikatza) [inline sbndcode/OpDetSim/wvfAna_module.cc:170]: curly brace on new line

File: sbndcode/OpDetSim/wvfana.fcl [hunk 7]
```diff
@@ -11,6 +11,8 @@ wvf_ana:
   OpDetsToPlot: ["pmt_coated", "pmt_uncoated",
                  "xarapuca_vuv", "xarapuca_vis",
                  "arapuca_vuv", "arapuca_vis"]
```
Comment (ikatza) [inline sbndcode/OpDetSim/wvfana.fcl:13]: kill `"arapuca_vuv", "arapuca_vis"` please

File: sbndcode/OpDetSim/opDetDigitizerWorker.cc [hunk 8]
```diff
@@ -186,31 +187,33 @@ void opdet::opDetDigitizerWorker::MakeWaveforms(opdet::DigiPMTSBNDAlg *pmtDigiti
         // getting only xarapuca channels with appropriate type of light
         else if((pdtype == "xarapuca_vuv" && !Reflected) ||
                 (pdtype == "xarapuca_vis" && Reflected) ) {
+          const bool is_daphne= fConfig.pdsMap.isElectronics(ch,"daphne");
+			    if(is_daphne){
```
Comment (ikatza) [inline sbndcode/OpDetSim/opDetDigitizerWorker.cc:191]: indentation

File: sbndcode/OpDetSim/opDetDigitizerWorker.cc [hunk 9]
```diff
@@ -186,31 +187,33 @@ void opdet::opDetDigitizerWorker::MakeWaveforms(opdet::DigiPMTSBNDAlg *pmtDigiti
...
+                                                  fConfig.Nsamples_Daphne);
           // including pre trigger window and transit time
           fWaveforms->at(ch) = raw::OpDetWaveform(fConfig.EnableWindow[0],
                                                   (unsigned int)ch,
                                                   waveform);
-        }
-        // getting only arapuca channels with appropriate type of light
-        else if((pdtype == "arapuca_vuv" && !Reflected) ||
-                (pdtype == "arapuca_vis" && Reflected) ) {
+            }else{
```
Comment (ikatza) [inline sbndcode/OpDetSim/opDetDigitizerWorker.cc:203]: new line for `else`

File: sbndcode/OpDetSim/opDetDigitizerWorker.cc [hunk 10]
```diff
@@ -265,34 +269,35 @@ void opdet::opDetDigitizerWorker::MakeWaveforms(opdet::DigiPMTSBNDAlg *pmtDigiti
...
           // including pre trigger window and transit time
           fWaveforms->at(ch) = raw::OpDetWaveform(fConfig.EnableWindow[0],
                                                   (unsigned int)ch,
                                                   waveform);
-        }
-        // getting only arapuca channels with appropriate type of light
-        if((pdtype == "xarapuca_vuv" && !Reflected) ||
-           (pdtype == "xarapuca_vis" && Reflected)) {
-          arapucaDigitizer->ConstructWaveform(ch,
+        }else{
```
Comment (ikatza) [inline sbndcode/OpDetSim/opDetDigitizerWorker.cc:287]: idem

File: sbndcode/OpDetSim/opDetDigitizerWorker.cc [hunk 11]
```diff
@@ -186,31 +187,33 @@ void opdet::opDetDigitizerWorker::MakeWaveforms(opdet::DigiPMTSBNDAlg *pmtDigiti
...
                                                   pdtype,
+                                                  is_daphne,
                                                   startTime,
                                                   fConfig.Nsamples);
           // including pre trigger window and transit time
           fWaveforms->at(ch) = raw::OpDetWaveform(fConfig.EnableWindow[0],
                                                   (unsigned int)ch,
                                                   waveform);
-      	}
+											  }
```
Comment (ikatza) [inline sbndcode/OpDetSim/opDetDigitizerWorker.cc:215]: indentation?

File: sbndcode/OpDetSim/ophitfinder_sbnd.fcl [hunk 12]
```diff
@@ -16,6 +16,8 @@ sbnd_hit_finder:
   PulsePolarityPMT:     -1         # use -1 for inverse polarity
   PulsePolarityArapuca:  1         # use -1 for inverse polarity
   UseDenoising:          true      # denoising algorithm to use with arapucas
+  DaphneFrequency: 					80.0  #in MHz. Frequency of the Daphne Readouts
```
Comment (ikatza) [inline sbndcode/OpDetSim/ophitfinder_sbnd.fcl:19]: align the value with others please

File: sbndcode/OpDetSim/opHitFinderSBND_module.cc [hunk 13]
```diff
@@ -85,16 +86,17 @@ namespace opdet {
...
     std::string opdetType;
+    std::string electronicsType;
     int threshold;
     std::vector<double> fwaveform;
     std::vector<double> outwvform;
     //int fSize;
     //int fTimePMT;         //Start time of PMT signal
     //int fTimeMax;         //Time of maximum (minimum) PMT signal
-    void subtractBaseline(std::vector<double>& waveform, std::string pdtype, double& rms);
+    void subtractBaseline(std::vector<double>& waveform, std::string pdtype, std::string electronicsType,double& rms);
```
Comment (ikatza) [inline sbndcode/OpDetSim/opHitFinderSBND_module.cc:96]: missing a space

File: sbndcode/OpDetSim/opHitFinderSBND_module.cc [hunk 14]
```diff
@@ -121,6 +123,7 @@ namespace opdet {
 
     auto const clockData = art::ServiceHandle<detinfo::DetectorClocksService const>()->DataForJob();
     fSampling = clockData.OpticalClock().Frequency(); // MHz
+    fSampling_Daphne = p.get<double >("DaphneFrequency" );;
```
Comment (ikatza) [inline sbndcode/OpDetSim/opHitFinderSBND_module.cc:126]: extra spaces, extra `;`

File: sbndcode/OpDetSim/opHitFinderSBND_module.cc [hunk 15]
```diff
@@ -178,14 +179,11 @@ namespace opdet {
         fwaveform[i] = wvf[i];
       }
 
-      subtractBaseline(fwaveform, opdetType, rms);
+      subtractBaseline(fwaveform, opdetType,electronicsType, rms);
```
Comment (ikatza) [inline sbndcode/OpDetSim/opHitFinderSBND_module.cc:182]: missing space

File: sbndcode/OpDetSim/opHitFinderSBND_module.cc [hunk 16]
```diff
@@ -197,15 +195,15 @@ namespace opdet {
       }
 
       // TODO: pass rms to this function once that's sorted. ~icaza
-      while(findAndSuppressPeak(fwaveform, timebin, Area, amplitude, threshold, opdetType)){
-        time = wvf.TimeStamp() + (double)timebin / fSampling;
+      while(findAndSuppressPeak(fwaveform, timebin, Area, amplitude, threshold, opdetType,electronicsType)){
```
Comment (ikatza) [inline sbndcode/OpDetSim/opHitFinderSBND_module.cc:198]: missing space

File: sbndcode/OpDetSim/opHitFinderSBND_module.cc [hunk 17]
```diff
@@ -197,15 +195,15 @@ namespace opdet {
...
+        if(electronicsType == "daphne"){
+        time = wvf.TimeStamp() + (double)timebin / fSampling_Daphne;}
+        else{
+        time = wvf.TimeStamp() + (double)timebin / fSampling;}
```
Comment (ikatza) [inline sbndcode/OpDetSim/opHitFinderSBND_module.cc:202]: better do:

```
if(electronicsType == "Daphne") time = wvf.TimeStamp() + (double)timebin / fSampling_Daphne;
else time = wvf.TimeStamp() + (double)timebin / fSampling;
```

or fix the indentation and add new lines for the curlies

File: sbndcode/OpDetSim/opHitFinderSBND_module.cc [hunk 18]
```diff
@@ -228,15 +226,17 @@ namespace opdet {
   DEFINE_ART_MODULE(opHitFinderSBND)
 
   void opHitFinderSBND::subtractBaseline(std::vector<double>& waveform,
-                                         std::string pdtype, double& rms)
+                                         std::string pdtype,std::string electronicsType, double& rms)
```
Comment (ikatza) [inline sbndcode/OpDetSim/opHitFinderSBND_module.cc:229]: missing space

File: sbndcode/OpDetSim/DigiArapucaSBNDAlg.hh [hunk 19]
```diff
@@ -60,9 +60,12 @@ namespace opdet {
       double XArapucaVISEff;    //XArapucaVIS efficiency (optical window + cavity)
       double DecayTXArapucaVIS;// Decay time of EJ280 in ns
       std::string ArapucaDataFile; //File containing timing structure for arapucas
+      bool SinglePEmodel; //Model for single pe response, false for ideal, true for test bench meas
 
       detinfo::LArProperties const* larProp = nullptr; ///< LarProperties service provider.
       double frequency; ///< Optical-clock frequency
+      double frequency_Daphne; ///< Optical-clock frequency for daphne readouts	
+
       CLHEP::HepRandomEngine* engine = nullptr;
```
Comment (ikatza) [inline sbndcode/OpDetSim/DigiArapucaSBNDAlg.hh:69]: I thought we agreed these could be `const`

Comment (ikatza) [inline sbndcode/OpDetSim/DigiArapucaSBNDAlg.hh:69]: where is `engine` used? below there's an `fEngine`

Comment (ikatza) [reply sbndcode/OpDetSim/DigiArapucaSBNDAlg.hh:69]: also above, please kill `ArapucaVUVEff` and `ArapucaVISEff`

File: sbndcode/OpDetSim/DigiArapucaSBNDAlg.hh [hunk 20]
```diff
@@ -80,58 +83,62 @@ namespace opdet {
...
 
-    double saturation;
+    const double fSampling;        //wave sampling frequency (GHz)
+    const double fSampling_Daphne;        //wave sampling frequency (GHz)
+    const double fXArapucaVUVEff;
+    const double fXArapucaVISEff;
+    const double fADCSaturation;
 
-    CLHEP::HepRandomEngine* fEngine; //!< Reference to art-managed random-number engine
+    CLHEP::HepRandomEngine* fEngine; //!< Reference to art-managed random-number engine //check this works ~rodrigoa
```
Comment (ikatza) [inline sbndcode/OpDetSim/DigiArapucaSBNDAlg.hh:108]: what needs to be checked? is it to do with the `const`?

perhaps a comment that can be removed?

File: sbndcode/OpDetSim/DigiArapucaSBNDAlg.cc [hunk 21]
```diff
@@ -41,25 +40,32 @@ namespace opdet {
...
+      std::vector<double>* SinglePEVec_40ftCable_Apsaia;
+      file->GetObject("SinglePEVec_40ftCable_Apsaia", SinglePEVec_40ftCable_Apsaia);
+      file->GetObject("SinglePEVec_40ftCable_Daphne", SinglePEVec_40ftCable_Daphne);
+      fWaveformSP = *SinglePEVec_40ftCable_Apsaia;
+      fWaveformSP_Daphne = *SinglePEVec_40ftCable_Daphne;}	
+  	else{
+	  	mf::LogDebug("DigiArapucaSBNDAlg") << " using ideal pe response";
+	    Pulse1PE(fWaveformSP,fSampling);
+		  Pulse1PE(fWaveformSP_Daphne,fSampling_Daphne);
+		  }
```
Comment (ikatza) [inline sbndcode/OpDetSim/DigiArapucaSBNDAlg.cc:67]: indentation

File: sbndcode/OpDetSim/DigiArapucaSBNDAlg.cc [hunk 22]
```diff
@@ -99,62 +107,39 @@ namespace opdet {
...
-          tphoton = (fTimeXArapucaVUV->fire());
-          tphoton += simphotons[i].Time - t_min;
+          double tphoton = (fTimeXArapucaVUV->fire()) + simphotons[i].Time - t_min;
           if(tphoton < 0.) continue; // discard if it didn't made it to the acquisition
           if(fParams.CrossTalk > 0.0 && (CLHEP::RandFlat::shoot(fEngine, 1.0)) < fParams.CrossTalk) nCT = 2;
           else nCT = 1;
-          timeBin = std::floor(tphoton * fSampling);
-          if(timeBin < wave.size()) AddSPE(timeBin, wave, nCT);
+          size_t timeBin = (is_daphne) ? std::floor(tphoton * fSampling_Daphne) : std::floor(tphoton * fSampling);
+          if(timeBin < wave.size()) {
```
Comment (ikatza) [inline sbndcode/OpDetSim/DigiArapucaSBNDAlg.cc:122]: maybe a GitHub rendering, but the lines don't seem aligned

File: sbndcode/OpDetSim/DigiArapucaSBNDAlg.cc [hunk 23]
```diff
@@ -99,62 +107,39 @@ namespace opdet {
...
+						if (!is_daphne) {AddSPE(timeBin, wave, fWaveformSP, nCT);
+						}
+            else{ AddSPE(timeBin, wave, fWaveformSP_Daphne, nCT);}
+          }
```
Comment (ikatza) [inline sbndcode/OpDetSim/DigiArapucaSBNDAlg.cc:142]: idem, alignment

File: sbndcode/OpDetSim/DigiArapucaSBNDAlg.cc [hunk 24]
```diff
@@ -253,46 +236,49 @@ namespace opdet {
...
-        wsp[i] = constT1 * std::exp(-(time - fParams.PeakTime) / fParams.FallTime);
-    }
+		double constT1 = fParams.ADC * fParams.MeanAmplitude;
+		for(size_t i = 0; i<fWaveformSP.size(); i++) {
+		  double time = static_cast<double>(i) / sampling;
+		  if (time < fParams.PeakTime)
+			fWaveformSP[i] = constT1 * std::exp((time - fParams.PeakTime) / fParams.RiseTime);
+		  else
+			fWaveformSP[i] = constT1 * std::exp(-(time - fParams.PeakTime) / fParams.FallTime);
+		}
```
Comment (ikatza) [inline sbndcode/OpDetSim/DigiArapucaSBNDAlg.cc:259]: indentation

File: sbndcode/OpDetSim/DigiArapucaSBNDAlg.cc [hunk 25]
```diff
@@ -378,7 +363,8 @@ namespace opdet {
     fBaseConfig.PeakTime          = config.peakTime();
     fBaseConfig.DecayTXArapucaVIS = config.decayTXArapucaVIS();
     fBaseConfig.ArapucaDataFile   = config.arapucaDataFile();
-
+    fBaseConfig.SinglePEmodel  =  config.singlePEmodel();
+		fBaseConfig.frequency_Daphne= config.DaphneFrequency();
```
Comment (ikatza) [inline sbndcode/OpDetSim/DigiArapucaSBNDAlg.cc:367]: indentation

**Review by ikatza — CHANGES_REQUESTED** [review]
Missing is mostly indentation and styling issues.

File: sbndcode/OpDetSim/sbndPDMapAlg_tool.cc (hunk 1 above)
Comment (rodralva) [reply sbndcode/OpDetSim/sbndPDMapAlg_tool.cc:33]: Removed the second comment

File: sbndcode/OpDetSim/sbndPDMapAlg_tool.cc (hunk 2 above)
Comment (rodralva) [reply sbndcode/OpDetSim/sbndPDMapAlg_tool.cc:39]: Removed the if

File: sbndcode/OpDetSim/wvfAna_module.cc (hunk 5 above)
Comment (rodralva) [reply sbndcode/OpDetSim/wvfAna_module.cc:169]: I believe is GitHub rendering 
image

File: sbndcode/OpDetSim/wvfAna_module.cc (hunk 6 above)
Comment (rodralva) [reply sbndcode/OpDetSim/wvfAna_module.cc:170]: Fixed
image

File: sbndcode/OpDetSim/opDetDigitizerWorker.cc (hunk 8 above)
Comment (rodralva) [reply sbndcode/OpDetSim/opDetDigitizerWorker.cc:191]: Again, I believe is github messing up tabs and spaces 
image

Comment (ikatza) [reply sbndcode/OpDetSim/opDetDigitizerWorker.cc:191]: So in that picture `arapucaDigitizer`  should start two spaces to the right. `litesimephotons, waveform, ...` should start even further to the right.

But whatevs, honestly future work is required in this module.

File: sbndcode/OpDetSim/DigiArapucaSBNDAlg.hh (hunk 19 above)
Comment (rodralva) [reply sbndcode/OpDetSim/DigiArapucaSBNDAlg.hh:69]: We made const the ones labeled with and f before them: 
image

Comment (rodralva) [reply sbndcode/OpDetSim/DigiArapucaSBNDAlg.hh:69]: It is the engine that initializes the fEngine in the .cc file:
image

File: sbndcode/OpDetSim/DigiArapucaSBNDAlg.hh (hunk 20 above)
Comment (rodralva) [reply sbndcode/OpDetSim/DigiArapucaSBNDAlg.hh:108]: Yes, seems like fEngine cant be made const:
image

File: sbndcode/OpDetSim/DigiArapucaSBNDAlg.cc (hunk 22 above)
Comment (rodralva) [reply sbndcode/OpDetSim/DigiArapucaSBNDAlg.cc:122]: I believe they are 
image

**Comment by pgreen135** [comment]
trigger build

**Review by ikatza — APPROVED** [review]
Thanks a lot for changing these @rodralva

**Comment by ikatza** [comment]
There should have been a warning from the reco stage, stating that there are less `OpHits`.

**Comment by pgreen135** [comment]
@rodralva @ikatza are the changes in the e20:prof CI tests above as expected? I am seeing small changes in the number of std::vector<raw::OpDetWaveform>, which seems to make sense.

@ikatza the reco stages use the reference detsim files as input, so wouldn't be affected by these changes in the detsim stage

**Comment by ikatza** [comment]
@rodralva should comment on this as he's the expert.

In this case I feel like this highlights a case were the CI as it is rather insensitive. Instead of the number of waveform we would like to know the size of these. Or as I was expecting how these changes affect further down the line.

**Comment by rodralva** [comment]
The number of waveforms remains the same. What I have seen is that the size of the output waveforms decreases notably. For example, running the wvf analizer to dump the digitizer waveforms on single interaction muon events produces 1.8 mb output files per event with develop. This goes down to 1.2/1.3 mb with this PR(~30% reduction). As for the reco stage, we are losing resolution, as the time tick goes from 2ns to 12.5 ns, but the fall time of the Xarapucas is greater than 100ns so the difference should be small. The test still used the ideal response, no big differences are expected(the waveforms have wider binning, but the shape remains the same) .

**Comment by pgreen135** [comment]
@rodralva does that mean these changes in the CI tests are not expected then? e.g. in the singles test:

`397: < DetSim | opdaq |  | std::vector<raw::OpDetWaveform> | 253
398: ---
399: > DetSim | opdaq |  | std::vector<raw::OpDetWaveform> | 245`

If so, could you investigate further, thanks.

**Pull request merged**